### PR TITLE
Use one thread for ACLK synchonization

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
Right now the agent will start multiple threads (one per host) to do the agent - cloud synchronization. This PR will only use one thread to handle the communication for all the the hosts and cut down resource usage. 

- [x] Use one thread to handle all ACLK communication to the cloud if the agent is cloud enabled and claimed
- [x] Correctly register the ACLK sync and Context services for a more reliable agent shutdown
- [x] Schedule parallel context loading for children to speed up startup time
       - When a child attempts to connect when context load is in progress the connection is temporarily rejected
       - Node update info to the cloud is also deferred until the context load is completed
- [x] Better cleanup of the chart labels metadata

### CODE REVIEW

**Comment by thiagoftsm** [comment]
Hello @stelfrag , please, take a look in [codacy](https://app.codacy.com/gh/netdata/netdata/pullRequest?prid=10958115).

**Comment by thiagoftsm** [comment]
@stelfrag why do we have an empty file named [database/rrdcontext.c](https://github.com/netdata/netdata/pull/14281/files#diff-06d0bd9290c25de90cda4b0f2cfc6b1ce8c8937fcdfba7927a5026895fefe9e7)?

**Comment by stelfrag** [comment]
> Hello @stelfrag , please, take a look in [codacy](https://app.codacy.com/gh/netdata/netdata/pullRequest?prid=10958115).

Will fix

**Comment by stelfrag** [comment]
> @stelfrag why do we have an empty file named [database/rrdcontext.c](https://github.com/netdata/netdata/pull/14281/files#diff-06d0bd9290c25de90cda4b0f2cfc6b1ce8c8937fcdfba7927a5026895fefe9e7)?

Ahh I will remove it -- it has been moved under contexts and with rebasing I added it empty by accident

File: database/sqlite/sqlite_aclk.c [hunk 1]
```diff
@@ -5,76 +5,174 @@
...
+        , guid
+        , (const char *) argv[IDX_OS]
+        , (const char *) argv[IDX_TIMEZONE]
+        , (const char *) argv[IDX_ABBREV_TIMEZONE]
+        , (int32_t) (argv[IDX_UTC_OFFSET] ? str2uint32_t(argv[IDX_UTC_OFFSET], NULL) : 0)
+        , (const char *) argv[IDX_TAGS]
+        , (const char *) (argv[IDX_PROGRAM_NAME] ? argv[IDX_PROGRAM_NAME] : "unknown")
+        , (const char *) (argv[IDX_PROGRAM_VERSION] ? argv[IDX_PROGRAM_VERSION] : "unknown")
+        , argv[3] ? str2i(argv[IDX_UPDATE_EVERY]) : 1
+        , argv[13] ? str2i(argv[IDX_ENTRIES]) : 0
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_aclk.c:119]: Considering what I am reading here right now, I suppose that `argv[3]` is exactly the same value than `argv[IDX_UPDATE_EVERY]`. The same logic is applied to `argv[13]`, to let this more clear and without any doubt, please, change the "magic number" for enum values.

File: database/sqlite/sqlite_aclk.c [hunk 2]
```diff
@@ -107,15 +205,13 @@ static int is_host_available(uuid_t *host_id)
 }
 
 // OPCODE: ACLK_DATABASE_DELETE_HOST
-void sql_delete_aclk_table_list(struct aclk_database_worker_config *wc, struct aclk_database_cmd cmd)
+static void sql_delete_aclk_table_list(char *host_guid)
 {
-    UNUSED(wc);
     char uuid_str[GUID_LEN + 1];
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_aclk.c:210]: You did not touch this code, but to improve code  I suggest to change `GUID_LEN +1` to `UUID_STR_LEN`.

File: database/sqlite/sqlite_aclk.c [hunk 3]
```diff
@@ -157,274 +253,66 @@ void sql_delete_aclk_table_list(struct aclk_database_worker_config *wc, struct a
...
-
-    fatal_assert(queue_size < ACLK_DATABASE_CMD_Q_MAX_SIZE);
-    /* enqueue command */
-    wc->cmd_queue.cmd_array[wc->cmd_queue.tail] = *cmd;
-    wc->cmd_queue.tail = wc->cmd_queue.tail != ACLK_DATABASE_CMD_Q_MAX_SIZE - 1 ?
-                         wc->cmd_queue.tail + 1 : 0;
-    wc->queue_size = queue_size + 1;
-    uv_mutex_unlock(&wc->cmd_mutex);
+    char sql[512];
+    snprintf(sql,511, "DELETE FROM aclk_alert_%s WHERE date_submitted IS NOT NULL AND "
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_aclk.c:292]: On line `275` (#define SQL_SELECT_ACLK_ACTIVE_LIST) and `300` (`#define SQL_SELECT_ACLK_ALERT_LIST`) we have sql queries as `define`, but here it is given direct as argument. The biggest difference is how we are passing arguments using strings (`%` in the others and `%s` and `%d` here). Are you sure we cannot move this `DELETE` for a define to keep one unique pattern in the code?

File: database/sqlite/sqlite_aclk.c [hunk 4]
```diff
@@ -594,206 +408,216 @@ static void aclk_database_worker(void *arg)
...
+    if (unlikely(!db_meta)) {
+        if (default_rrd_memory_mode != RRD_MEMORY_MODE_DBENGINE) {
+            return;
+        }
+        error_report("Database has not been initialized");
+        return;
+    }
+
+    info("Creating archived hosts");
+    rc = sqlite3_exec_monitored(db_meta, "SELECT host_id, hostname, registry_hostname, update_every, os, "
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_aclk.c:537]: Like one of previous comments, please, move this `SELECT` to a `define`.

File: database/sqlite/sqlite_aclk.c [hunk 5]
```diff
@@ -594,206 +408,216 @@ static void aclk_database_worker(void *arg)
...
+
+    if (rc != SQLITE_OK) {
+        error_report("SQLite error when loading archived hosts, rc = %d (%s)", rc, err_msg);
+        sqlite3_free(err_msg);
+    }
+    // Trigger host context load for hosts that have been created
+    metadata_queue_load_host_context(NULL);
+
+#ifdef ENABLE_ACLK
+    rc = sqlite3_exec_monitored(db_meta, "SELECT ni.host_id, ni.node_id FROM host h, node_instance ni " \
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_aclk.c:550]: Another one that can be moved is here.

File: database/sqlite/sqlite_aclk.c [hunk 6]
```diff
@@ -594,206 +408,216 @@ static void aclk_database_worker(void *arg)
...
+
+    queue_aclk_sync_cmd(ACLK_DATABASE_PUSH_ALERT_CONFIG, node_id, config_hash);
+}
+
+void aclk_push_node_alert_snapshot(const char *node_id)
+{
+    if (unlikely(!aclk_sync_config.initialized))
+        return;
+
+    queue_aclk_sync_cmd(ACLK_DATABASE_PUSH_ALERT_SNAPSHOT, strdupz(node_id), NULL);
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_aclk.c:590]: Are we not having memory leaks here? I remember to chat with @MrZammler about this one day, but I am not sure if it was with the same function.

Comment (thiagoftsm) [reply database/sqlite/sqlite_aclk.c:590]: If this is happening with this function, it is also happening in the nexts.

File: database/sqlite/sqlite_aclk.c (hunk 3 above)
Comment (thiagoftsm) [reply database/sqlite/sqlite_aclk.c:292]: Another change here, please, use `snprintfz` instead `snprintf`.

File: database/sqlite/sqlite_aclk_alert.c [hunk 7]
```diff
@@ -747,100 +748,95 @@ void aclk_start_alert_streaming(char *node_id, uint64_t batch_id, uint64_t start
...
 
-    db_execute(buffer_tostring(sql));
+    if (unlikely(!host || !(wc = host->aclk_sync_host_config)))
+        return;
 
-    log_access("ACLK STA [%s (%s)]: QUEUED REMOVED ALERTS", wc->node_id, wc->host ? rrdhost_hostname(wc->host) : "N/A");
+    char sql[ACLK_SYNC_QUERY_SIZE * 2];
 
-    buffer_free(sql);
+    snprintfz(sql,ACLK_SYNC_QUERY_SIZE * 2 - 1, "INSERT INTO aclk_alert_%s (alert_unique_id, date_created, filtered_alert_unique_id) " \
```
Comment (thiagoftsm) [inline database/sqlite/sqlite_aclk_alert.c:791]: Here we have another query given directly as argument.

Comment (thiagoftsm) [reply database/sqlite/sqlite_aclk_alert.c:791]: We also have more queries inside this file, so if you agree to move all them to `define`, please, take a look in other arguments inside this file.

File: database/sqlite/sqlite_aclk.c (hunk 6 above)
Comment (stelfrag) [reply database/sqlite/sqlite_aclk.c:590]: > 

Good catch.. there is a memory leak under certain condition. I will fix

**Review by thiagoftsm — DISMISSED** [review]
Considering the received [answer](https://github.com/netdata/netdata/pull/14281/files#r1132645160),  and tests that I ran during one hour, LGTM!

**Review by thiagoftsm — DISMISSED** [review]
I retested after last commits and rebase, and I have parent and child exposed correctly on cloud, LGTM!

**Review by underhood — DISMISSED** [review]
quick scan and test, nothing scary detected

**Review by MrZammler — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]
I finish another round of tests and everything worked as expected.

**Pull request merged**